### PR TITLE
Simplify extension setup.

### LINKED ISSUE #6928: Cannot run `setup.py build` with numpy master
Since numpy/numpy#7853 landed in master, running `python setup.py build` for matplotlib will now fail during the `build_ext` section:

```
running build_ext
Traceback (most recent call last):
  File "setup.py", line 304, in <module>
    **extra_args
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "setup.py", line 147, in run
    return BuildExtCommand.run(self)
  File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_ext.py", line 66, in run
    _build_ext.run(self)
  File "/usr/local/lib/python2.7/dist-packages/Cython/Distutils/build_ext.py", line 164, in run
    _build_ext.build_ext.run(self)
  File "/usr/lib/python2.7/distutils/command/build_ext.py", line 337, in run
    self.build_extensions()
  File "/usr/local/lib/python2.7/dist-packages/Cython/Distutils/build_ext.py", line 171, in build_extensions
    ext.sources = self.cython_sources(ext.sources, ext)
  File "/usr/local/lib/python2.7/dist-packages/Cython/Distutils/build_ext.py", line 242, in cython_sources
    extension.include_dirs = list(extension.include_dirs)
  File "~/matplotlib/setupext.py", line 910, in __get__
    result = obj._hooks[self._name]() + result
  File "~/matplotlib/setupext.py", line 937, in include_dirs_hook
    reload(numpy)
  File "/usr/local/lib/python2.7/dist-packages/numpy-1.12.0.dev0+5130ef1-py2.7-linux-x86_64.egg/numpy/__init__.py", line 117, in <module>
    raise RuntimeError('Reloading numpy is not supported')
RuntimeError: Reloading numpy is not supported
```

It looks like the reload was introduced by https://github.com/matplotlib/matplotlib/pull/2445/commits/373c4d79dfc526fb91ca0d00407b6d244ca9dc62, but I'm not sure why it's needed or if it can be worked around.

### PR DESCRIPTION
**Pull request by anntzer**
setupext currently has a complex machinery to define extension modules.
Essentially, the problem is that numpy may not be installed at the time
the extensions are declared, so we can't call np.get_include() to get
the include paths.  So we use a DelayedExtension class and a hook
mechanism to inject the numpy include path later, once it becomes
available.

Instead, we can just declare a dummy extension (so that setuptools
doesn't elide the call to build_ext), then swap in the correct
extensions in build_ext.finalize_options(): at that point, numpy will
have been installed and we don't need any crazy machinery.

Also closes #6928 by not attempting to reload numpy anymore.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: setupext.py [hunk 1]
```diff
@@ -1211,7 +1131,14 @@ def get_extension(self):
         pkg_config.setup_extension(
             ext, 'libpng', default_libraries=['png', 'z'],
             alt_exec='libpng-config --ldflags')
-        Numpy().add_flags(ext)
+        # We call this twice: once early, to check whether png exists (from
```
Comment (anntzer) [inline setupext.py]: this hack won't be necessary after #11983 goes in so let's wait for that before rebasing this PR

**Comment by anntzer** [comment]
Rebased on top of #11983 which makes things a bit simpler.

**Comment by tacaswell** [comment]
Tried to power cycled to see if it would simplify the diff now that #11983 is merged, @anntzer mind rebasing this again to make it easier to review?

**Comment by anntzer** [comment]
Done.

**Comment by jklymak** [comment]
Marking release critical so it gets some love..  I'd review, but really don't grok whats going on here.  But if someone more knowledgeable reviewed, I'd be comfortable approving as not doing anything dangerous.

**Comment by timhoffm** [comment]
Sorry, can’t help either. Not experienced in setuptools.

**Comment by anntzer** [comment]
@tacaswell do you want to entice a second reviewer? :)

File: setup.py [hunk 2]
```diff
@@ -101,6 +101,12 @@ def __init__(self, dist):
 
 
 class BuildExtraLibraries(BuildExtCommand):
+    def finalize_options(self):
+        self.distribution.ext_modules[:] = [
```
Comment (QuLogic) [inline setup.py]: Assigning to `[:]` will expand the input, so you don't actually need the `[*]` around `filter`.

Comment (anntzer) [reply setup.py]: good point, changed

**Comment by dstansby** [comment]
Going to merge based on two previous +ive reviews.

**Pull request merged**